mifospay/src/main/java/org/mifos/mobilewallet/mifospay/qr/ui/ReadQrActivity.kt

```diff
@@ -0,0 +1,139 @@
+    private lateinit var mPresenter: ReadQrPresenter
+    private lateinit var mReadQrPresenter: QrContract.ReadQrPresenter
+    private lateinit var binding: ActivityReadQrBinding
+
+    public override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        binding = ActivityReadQrBinding.inflate(layoutInflater)
+        val view = binding.root
+        setContentView(view)
+
+        setToolbarTitle(Constants.SCAN_CODE)
+        showColoredBackButton(Constants.BLACK_BACK_BUTTON)
+        mPresenter.attachView(this)
+        binding.scannerView.setAutoFocus(true)
+
+        binding.btnFlashOn.setOnClickListener {
+            turnOnFlash()
+        }
+
+        binding.btnFlashOff.setOnClickListener {
+            turnOffFlash()
+        }
+
+        binding.btnOpenGallery.setOnClickListener {
+            openGallery()
+        }
+
+    }
+
+    fun turnOnFlash() {
+        binding.scannerView.flash = true
+        binding.btnFlashOn.visibility = View.GONE
+        binding.btnFlashOff.visibility = View.VISIBLE
+    }
+
+    fun turnOffFlash() {
+        binding.scannerView.flash = false
+        binding.btnFlashOn.visibility = View.VISIBLE
+        binding.btnFlashOff.visibility = View.GONE
+    }
+
+    fun openGallery() {
+        val photoPic = Intent(Intent.ACTION_PICK)
+        photoPic.type = "image/*"
+        startActivityForResult(photoPic, SELECT_PHOTO)
+    }
+
+    public override fun onResume() {
+        super.onResume()
+        binding.scannerView.setResultHandler(this)
+        binding.scannerView.startCamera()
+    }
+
+    public override fun onPause() {
+        super.onPause()
+        binding.scannerView.stopCamera()
+    }
+
+    override fun handleResult(result: Result) {
+        val qrData = result.text
+        val returnIntent = Intent()
+        returnIntent.putExtra(Constants.QR_DATA, qrData)
+        setResult(RESULT_OK, returnIntent)
+        finish()
+    }
+
+    override fun setPresenter(presenter: QrContract.ReadQrPresenter?) {
+        if (presenter != null) {
+            mReadQrPresenter = presenter
+        }
+    }
+
+    private fun scanQRImage(bMap: Bitmap): String? {
+        var contents: String? = null
+        val intArray = IntArray(bMap.width * bMap.height)
+        bMap.getPixels(intArray, 0, bMap.width, 0, 0, bMap.width, bMap.height)
+        val source: LuminanceSource = RGBLuminanceSource(bMap.width, bMap.height, intArray)
+        val bitmap = BinaryBitmap(HybridBinarizer(source))
+        val reader: Reader = MultiFormatReader()
+        try {
+            val result = reader.decode(bitmap)
+            contents = result.text
+            val returnIntent = Intent()
+            returnIntent.putExtra(Constants.QR_DATA, contents)
+            setResult(RESULT_OK, returnIntent)
+            finish()
+        } catch (e: Exception) {
+            Toast.makeText(
+                applicationContext,
```

Comment:
No need to use application context here. You can go through this [article](https://blog.mindorks.com/understanding-context-in-android-application-330913e32514) to learn more

Reply:
Thanks for the article